### DISCUSSION TITLE
sqlite extensions (e.g. sqlite-zstd) support

### DISCUSSION THREAD

**Discussion by allquixotic** [SeaQL/sea-orm#1432] [Ideas]
It would be neat if sea-orm could support [sqlite-zstd](https://github.com/phiresky/sqlite-zstd). This library is one of perhaps several "sqlite extension libraries" (I'm not sure exactly how that extension mechanism works; it might be extensibility deliberately provided by libsqlite3, or some kind of override). 

The basic idea of a sqlite extension is that it's mostly transparent to any C or Rust layers calling into libsqlite3, but it has to be loaded first. Then, optionally, the extension might enable new SQL syntax. Since sea-orm already seems to expose a mechanism to issue raw SQL to the DB layer, sqlite-zstd's requirement for the user to enable row-level compression using SQL statements should not be a major barrier to the user.

The main barrier as I understand it is that sea-orm uses its own libsqlite3 dependency without supporting an extensibility mechanism.

I know that other extension libraries exist for sqlite that add other features, but row-level compression is probably one of the most transparent and easiest to implement. Anything that really messes with data types or semantics of existing SQL statements would probably not work with the ORM's generated SQL. Still, most compression extensions should be transparent.

Goals of making this into a feature would entail:

 - What general approach should be pursued to add extension support?
 - Should extension support be generic, in a way that the user can add any extensions they want? Or should it be for specific libraries with specific requirements?
 - Should extension support be "early-bound" with support for specific extensions directly available upon request in the sea-orm API, or should it be "late-bound" with the user requesting an extension by filename and having sea-orm look for it at runtime?
 - Can extension support for sqlite-zstd and similar be made specifically easy for sqlite-zstd without adding a crate dependency on sqlite-zstd as a rust library, or do we have to pull in the crate at build time?  Or maybe adding the crate is _desirable_?

One of the nice things about sqlite-zstd is that it's written in Rust, and has a Rust library as part of the solution, so it's probably in a fairly unique class of software in terms of what it provides to users/ORM layers. If you designed the extension mechanism to just look for a native shared library (dll/so/dylib) to dynamically load, you would not be getting the benefit of directly hooking into the Rust crate that sqlite-zstd provides (that benefit may include, for example, improved memory safety).

**Comment by billy1624** [comment]
Hey @allquixotic, sorry for the delay. I think SQLx already support loading SQLite extension?
https://docs.rs/sqlx/latest/sqlx/sqlite/struct.SqliteConnectOptions.html#method.extension
